File: typescript/euler/bin/p0017.ts

// project euler: problem 17

import { range } from "../lib/util.ts";

const words = new Map<number, number>([
  [1, "one".length],
  [2, "two".length],
  [3, "three".length],
  [4, "four".length],
  [5, "five".length],
  [6, "six".length],
  [7, "seven".length],
  [8, "eight".length],
  [9, "nine".length],
  [10, "ten".length],
  [11, "eleven".length],
  [12, "twelve".length],
  [13, "thirteen".length],
  [14, "fourteen".length],
  [15, "fifteen".length],
  [16, "sixteen".length],
  [17, "seventeen".length],
  [18, "eighteen".length],
  [19, "nineteen".length],
  [20, "twenty".length],
  [30, "thirty".length],
  [40, "forty".length],
  [50, "fifty".length],
  [60, "sixty".length],
  [70, "seventy".length],
  [80, "eighty".length],
  [90, "ninety".length],
  [0, 0], // Data when remainder is 0
]);

export function compute(limit: number): string {
  const trunc = Math.trunc;

  let acc = 0;
  for (const n of range(1, limit + 1)) {
    if (n === 1000) {
      acc += 11;
    } else if (n < 20) {
      acc += words.get(n)!;
    } else if (n < 100) {
      acc += words.get(n - (n % 10))! + words.get(n % 10)!;
    } else if (n % 100 === 0) {
      /* "xxx" hundred -> "xxx".length + 7 */
      acc += words.get(trunc(n / 100))! + 7;
    } else if (n % 100 < 20) {
      /* "xxx" hundred and ... -> "xxx".length + 7 + 3 + ... */
      acc += words.get(trunc(n / 100))! + 7 + 3 + words.get(n % 100)!;
    } else {
      /* "xxx" hundred and ... -> "xxx".length + 7 + 3 + ... */
      acc += words.get(trunc(n / 100))! + 7 + 3 + words.get((n % 100) - (n % 10))! + words.get(n % 10)!;
    }
  }

  return String(acc);
}

export function solve(): void {
  const t0 = performance.now();
  const result = compute(1000);
  const t1 = performance.now();
  const duration_ms = (t1 - t0).toFixed(4);

  console.log(`Answer: ${result}`);
  console.log(`Elapsed time: ${duration_ms} msec.`);

  return;
}

File: python/euler/bin/p0017.py
# project euler: problem 17

words = dict(
    [
        (1, len('one')),
        (2, len('two')),
        (3, len('three')),
        (4, len('four')),
        (5, len('five')),
        (6, len('six')),
        (7, len('seven')),
        (8, len('eight')),
        (9, len('nine')),
        (10, len('ten')),
        (11, len('eleven')),
        (12, len('twelve')),
        (13, len('thirteen')),
        (14, len('fourteen')),
        (15, len('fifteen')),
        (16, len('sixteen')),
        (17, len('seventeen')),
        (18, len('eighteen')),
        (19, len('nineteen')),
        (20, len('twenty')),
        (30, len('thirty')),
        (40, len('forty')),
        (50, len('fifty')),
        (60, len('sixty')),
        (70, len('seventy')),
        (80, len('eighty')),
        (90, len('ninety')),
        (0, 0),  # special data
    ]
)


def compute(limit: int) -> str:
    cnt = 0
    for n in range(1, limit + 1):
        if n == 1000:
            # one thouthand (3 + 8)
            cnt += 11
        elif n < 20:
            cnt += words[n]
        elif n < 100:
            cnt += words[n - (n % 10)] + words[n % 10]
        elif n % 100 == 0:
            # xxx hundred (len(xxx) + 7)
            cnt += words[n // 100] + 7
        elif n % 100 < 20:
            # xxx hundred and ...
            cnt += words[n // 100] + 7 + 3 + words[n % 100]
        else:
            # xxx hundres and ...
            cnt += words[n // 100] + 7 + 3 + words[(n % 100) - (n % 10)] + words[n % 10]

    return str(cnt)


def solve() -> str:
    return compute(1000)
